Allowed users to remove the animation that happens when a ScrollView's extents shrink

gaaclarke opened the pull request:
## Description

As reported in the issue, shrinking a ScrollView such that the scroll offset is outside the bounds currently results in a ballistic animation to correct being out of bounds.  This PR allows users to remove that animation and just jumps the scroll position to be in bounds in one frame.

## Related Issues

https://github.com/flutter/flutter/issues/55610

## Tests

I added the following tests:

- scroll_position_test: 'shrink listview'
- scroll_position_test: 'shrink listview while dragging'

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*
   - [ ] I wrote a design doc: https://flutter.dev/go/template *Replace this with a link to your design doc's short link*
   - [ ] I got input from the developer relations team, specifically from: *Replace with the names of who gave advice*
   - [ ] I wrote a migration guide: *Replace with a link to your migration guide*

<!-- Links -->
[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

Hixie commented on packages/flutter/lib/src/widgets/scroll_position_with_single_context.dart:
```diff
@@ -17,6 +17,21 @@ import 'scroll_notification.dart';
+  @override
+  double x(double time) => destination;
+}
+
```
we should be able to just jump if we don't actually want to animate. saves us the resources of creating the animation, etc.

gaaclarke committed:
Started clamping the pixel offset to the scroll bounds if a boundary
change causes the scroll offset to be out of bounds.

gaaclarke commented:
@Hixie What do you think about this approach?  I took what you did and changed it in a few ways:
1) Don't delegate to physics since we want the same behavior regardless of physics and this keeps things simpler.
2) Force the activity to go idle.  Since we've done a warp we don't want to continue any other activity.

gaaclarke committed:
moved over ian's delegating to scrollphysics

gaaclarke committed:
added a test for idleactivity

gaaclarke committed:
made the test a little less fragile by specifying a screen size

gaaclarke committed:
Revert "made the test a little less fragile by specifying a screen size"

This reverts commit becc3dbb9bec61d8f0f7f8177e08eb2e45915d76.

gaaclarke committed:
cleaned up tests, added one for resizing while dragging

gaaclarke committed:
started handling collapsing while dragging.

gaaclarke committed:
update docstring

gaaclarke committed:
Turned off the feature for NestedScrollViews.  When playing with them
they looked fine but the tests fail.

gaaclarke committed:
fixed some list wheel scroll view tests, they were starting a scroll
controller out of bounds, i'm breaking that

gaaclarke committed:
adjusted the clamp_overscrolls_test

Hixie commented on packages/flutter/lib/src/widgets/nested_scroll_view.dart:
```diff
@@ -1182,6 +1182,11 @@ class _NestedScrollPosition extends ScrollPosition implements ScrollActivityDele
+  @override
+  bool doesAcceptAdjustedPositionForNewDimensions(double adjustedPixels) {
+    return false;
+  }
```
assuming doesAcceptAdjustedPositionForNewDimensions exists just for this, could we instead just have the NestedScrollView logic pass in a ScrollPhysics class that overrides adjustPositionForNewDimensions?

gaaclarke replied:
I was following your lead, you had a similar method on ScrollPosition.  I thought your intention was to allow people to customize there as well.  I can revisit this once I get the baseline functionality.

Hixie commented on packages/flutter/lib/src/widgets/scroll_physics.dart:
```diff
@@ -228,6 +228,50 @@ class ScrollPhysics {
+  }) {
+    return (parent == null)
+      ? position.pixels
+      : parent.adjustPositionForNewDimensions(position, isScrolling: isScrolling, velocity: velocity);
```
we should use the style that is used in the other functions here, it's more readable than ?: imho (or rather, it more accurately represents the prototype inheritance that is being implemented -- it's not like we're considering two different alternatives, rather we're returning the default if there's no parent, and otherwise we're falling back on the parent).

gaaclarke replied:
It sounds like you attach semantics to the ternary operator that aren't necessarily associated with it.  Usage of the ternary operator can help with 2 things:
1) Reduce mutations:
```
int value;
if (isTrue) {
  value = 1;
} else {
  value = 0;
}
```
2) Reduce return statements in a procedure:
```
int foo() {
  if (isTrue) {
    return 1;
  }
  return 0;
}
```

Both mutations and multiple exits on a procedure are generally regarded as things to be reduced to clean up code, so the ternary operator can help us there.

Hixie replied:
mostly I'm just eager to keep the style in this file consistent.

gaaclarke replied:
done

Hixie commented on packages/flutter/lib/src/widgets/scroll_physics.dart:
```diff
@@ -500,6 +552,14 @@ class ClampingScrollPhysics extends ScrollPhysics {
+    @required bool isScrolling,
+    @required double velocity,
+  }) {
+    return position.pixels.clamp(position.minScrollExtent, position.maxScrollExtent) as double;
```
when I was noodling on this I had assumed that if we do it even when velocity is zero, it will cause animations to flail because the animation will immediately try to snap it back... is that not the case?

gaaclarke replied:
The system is complicated enough that I'm trying to move to being test driven and ripped out assumptions.  Those needed to be ripped out specifically to get dragging working.  I can add a test for shrinking the view while it's ballistic.

gaaclarke replied:
I was just playing around with the stepper.  Looks like if you animate the resizing of you scrollview this will just poke the animation to make sure it stays in bounds.  Jumping might not be erroneous then if you abruptly change the size, if you want it smooth, use an animation.

Hixie commented on packages/flutter/lib/src/widgets/scroll_physics.dart:
```diff
@@ -398,6 +442,14 @@ class BouncingScrollPhysics extends ScrollPhysics {
+    @required bool isScrolling,
+    @required double velocity,
+  }) {
+    return position.pixels.clamp(position.minScrollExtent, position.maxScrollExtent) as double;
```
this is the main thing from my patch that is going to need more work -- specifically, we need to handle the case of being overscrolled in a more graceful manner.

Ideally what I'd love for us to do is if we are under/over scrolled, we maintain the delta from the min/max extent to the position before and after the correction. The difficulty is that here we're not passing in the old min/max extent. Maybe we could just pass them in? I guess we have them in applyContentDimensions which is what is calling this...

cc @goderbauer

Hixie replied:
maybe we pass in two ScrollMetrics, one from before the update and one from after?
If we do this then we should probably avoid calling this method in the case where the old metrics were null...

gaaclarke replied:
Check out my widget test for resizing while dragging.  I think that captures what we were saying yesterday about keeping a valid drag going.

gaaclarke committed:
adjusted another  test

gaaclarke committed:
fixed another test

gaaclarke committed:
fixed another test

gaaclarke committed:
Revert "fixed another test"

This reverts commit 318e0645f30492d1f8ea9ac7f8faaaed04860d25.

gaaclarke committed:
- gave up on adjusting the drag gesture
- started supporting old tests by not correcting initialPosition
- started filtering out corrections that happen to scroll views that dabble in infinity

gaaclarke committed:
added  animation assert

gaaclarke committed:
undid changes to stepper test

gaaclarke committed:
turned of bounds clamping in AlwaysScrollableScrollPhysics

gaaclarke committed:
revert last change

gaaclarke committed:
undid most of the work, gave up on changing the default behavior and
instead made sure that someone can create the behavior if they want.

gaaclarke committed:
revert test change

gaaclarke committed:
Merge branch 'master' into remove-list-extents-animation

gaaclarke commented on packages/flutter/lib/src/widgets/scroll_position.dart:
```diff
@@ -489,11 +489,20 @@ abstract class ScrollPosition extends ViewportOffset with ScrollMetrics {
       _haveDimensions = true;
-      applyNewDimensions();
       _didChangeViewportDimensionOrReceiveCorrection = false;
+      if (!isFirstApplication) {
```
In the course of trying to get this working there are a dozen tests that assume you can start a ScrollPosition out of bounds.  I think it has no ramifications in practical use, but keeps tests from breakinig.

gaaclarke committed:
removed ternary operator

Hixie commented on packages/flutter/lib/src/widgets/scroll_position.dart:
```diff
@@ -489,11 +489,20 @@ abstract class ScrollPosition extends ViewportOffset with ScrollMetrics {
       _minScrollExtent = minScrollExtent;
       _maxScrollExtent = maxScrollExtent;
       _haveDimensions = true;
-      applyNewDimensions();
```
@gaaclarke was there a particular thing that led to you inlining applyNewDimensions here, as opposed to letting people override it?

Hixie replied:
er, i mean `approveNewDimensions`

Hixie replied:
ah, i see -- that was a response to my comment above.

Hixie commented:
(i'm going to close this PR and upload a new one -- thanks for the tests @gaaclarke!)

gaaclarke commented:
@Hixie No problem, are you uploading a new PR because you believe you can change the default behavior without breaking all ScrollViews as opposed to just letting people customize the behavior?  I'm pretty confident it can't be done but I'm also not as familiar with the codebase.  I'd be happy to be proven wrong on that point.

Hixie commented:
that's my goal (might involve a massive rewrite, we'll see).

Hixie commented:
Updated PR: https://github.com/flutter/flutter/pull/56521

Turns out most of the test breakage was an off-by-one error in the fixed extent sliver layout logic.

Hixie commented on packages/flutter/test/widgets/scroll_position_test.dart:
```diff
@@ -140,6 +140,75 @@ Future<void> performTest(WidgetTester tester, bool maintainState) async {
+}
+
+class _ExpandingBoxState extends State<ExpandingBox> with AutomaticKeepAliveClientMixin<ExpandingBox>{
+  _ExpandingBoxState({this.collapsedSize, this.expandedsize}) {
```
the idiomatic thing here btw is to use `widget.collapsedSize` and for the `State` subclass to not have any constructor arguments.

Hixie commented on packages/flutter/test/widgets/scroll_position_test.dart:
```diff
@@ -160,6 +229,110 @@ void main() {
+
+    final ScrollPosition position =
+        tester.state<ScrollableState>(find.byType(Scrollable)).position;
+    expect(position.activity.runtimeType, IdleScrollActivity);
```
fwiw, the idiomatic thing here is to use `isInstanceOf<T>()` as the matcher, rather than use `runtimeType`.

Hixie commented on packages/flutter/test/widgets/scroll_position_test.dart:
```diff
@@ -160,6 +229,110 @@ void main() {
+    await tester.pumpAndSettle();
+    await drag1.up();
+    await tester.pumpAndSettle();
+    expect(position.pixels, closeTo(500.00, 0.01));
```
fwiw, `moreOrLessEquals` is a more practical way of doing `closeTo` when you are just trying to side-step floating point issues. (It doesn't require the second argument and will in due course be updated to actually do it mathematically correctly for floating point.)

Hixie commented on packages/flutter/test/widgets/scroll_position_test.dart:
```diff
@@ -160,6 +229,110 @@ void main() {
+    await drag2.up();
+    await tester.pumpAndSettle();
+    expect(position.maxScrollExtent, 900.0);
+    expect(position.pixels, closeTo(900.00, 0.01));
```
why do you want this to be 900?

Hixie commented on packages/flutter/test/widgets/scroll_position_test.dart:
```diff
@@ -160,6 +229,110 @@ void main() {
+    expect(position.maxScrollExtent, 100.0);
+    expect(position.pixels, 0.0);
+    await tester.tap(find.byType(FlatButton));
+    await tester.pumpAndSettle();
```
you use `pumpAndSettle` a lot in this test; in general we should use `pump`, and `pumpAndSettle` is reserved for cases where we want to count the number of frames that an animation takes.
